### ISSUE TITLE
Exception for r.and when used with r.args

### ISSUE THREAD

**Issue by alexeygolev** [neumino/rethinkdbdash#115]
The following code works in _Data Explorer_
```javascript
r.db('db').table('table')
    .coerceTo('array')
    .filter(function(cP){
      return r.and(r.args(r.expr([{
          ids:['1bc07fca-2f68-4e2e-bc47-022b162f81a9', 'd4b99c17-eea8-4583-b34e-2a9dd6174f90'],
          all: true
        },
        {
          ids:['bc2232ee-ce92-4c4d-999e-c52719dd79ea'],
          all: false
        }]).map(
        function(rule) {
          return r.branch (
            rule('all'),
            r.and(r.args(rule('ids').map(function(id){
              return cP('tags').contains(id)
            }))),
            r.and(r.args(rule('ids').map(function(id){
              return cP('tags').contains(id)
            }))))
        }
      )))
 })

```

But throws this when used with rethinkdbdash
``` 
ReqlDriverError: `and` takes at least 1 argument, 0 provided after:
  r.args(var_2("ids").map(function(var_3) {
      return var_1("tags").contains(var_3)
  }))
```
`r.or` throws as well.
As far as I can understand from reading rethinkdbdash's source and tests, it throws the error for the _infix_ version of `and` (otherwise it would have thrown 'at least 2' according to the tests). Not sure why. In any case it doesn't seem to like `r.args` that much.

This is an ES6 version which is a bit easier to read
```javascript
  return r.db('db').table('table')
    .coerceTo('array')
    .filter(cP => r.and(r.args(r.expr(rules)
      .map(rule => r.branch(
        rule('all'),
        r.and(r.args(rule('ids').map(id => cP('tags').contains(id)))),
        r.or(r.args(rule('ids').map(id => cP('tags').contains(id))))
      )))))

```

**Comment by alexeygolev** [comment]
From what I can see in the source `r.and` calls its "infix" version internally.
https://github.com/neumino/rethinkdbdash/blob/master/lib/index.js#L244
The problem though that we are effectively ending up with `r.expr(r.args(...)).and(undefined)`.
This way when `Term.prototype.and` does the `_arityRange` here
https://github.com/neumino/rethinkdbdash/blob/master/lib/term.js#L1609
it throws an error.

**Comment by alexeygolev** [comment]
So a quick hack would be adding the following:
```javascript
  if(_args.length === 0 && this._query[0] === termTypes.ARGS) {
    var term = new Term(this._r, this._error);
    term._query.push(termTypes.AND)
    var args = [this._query];
    term._query.push(args);
    return term;
  }
```
after https://github.com/neumino/rethinkdbdash/blob/master/lib/term.js#L1608
Which obviously has to be abstracted away because it should be used in quite a few other functions (eg for OR in my use case). But I'm a bit hesitant to work on the pull request because it might be that this is too hacky and there is a proper way.

**Comment by neumino** [comment]
Thanks @alexeygolev for the report.
It should be fixed in 2.0.18
